relayer/src/client.rs

```diff
@@ -1,120 +1,296 @@
-
-                    if !receipts.is_empty() {
-                        let json = serde_json::to_string(&receipts)?;
-                        db.insert_receipts(block_hash, &json)?;
-                    } else {
-                        log::debug!(
-                            "Block {} does not have any receipts",
-                            block_hash.encode_hex()
-                        );
-                    }
-                } else {
-                    log::info!(
-                        "Could not get a block by block_hash {}",
-                        block_hash.encode_hex()
-                    );
-                    continue 'outer;
-                }
+            if let Err(e) = self.db.insert_or_update_finalized_block_info(
+                finalized_block.number,
+                H256(finalized_block.hash.0),
+            ) {
+                log::error!(target: TARGET,"Failed to update latest finalized block info: {}", e);
+                continue;
             }
+            latest_finalized_block = Some(finalized_block.number);
+            if let Err(e) = self
+                .collect_blocks_after_finality_update(finalized_block)
+                .await
+            {
+                log::error!(target: TARGET,"Failed to process finality update: {}", e);
+            } else {
+                log::info!(target: TARGET,"Processed finality update");
+            };
         }
     }
-}
 
-fn network(config: &Config) -> Network {
-    match &config.network {
-        Some(network) => {
-            if network == "goerli" {
-                Network::GOERLI
+    /// Fetches all blocks from the web3 provider. The fetching goes backwards from the latest finalized block
+    /// to the latest processed block using parent hash.
+    async fn collect_blocks_after_finality_update(
+        &mut self,
+        finalized_block: ExecutionBlock,
+    ) -> Result<()> {
+        const TARGET: &str = "relayer::client::collect_blocks_after_finality_update";
+
+        log::info!(target: TARGET,"Processing finality update");
+        let latest_processed_block = self
+            .db
+            .select_latest_block_height(BlockType::Processed)?
+            .unwrap_or(finalized_block.number - BLOCK_AMOUNT_TO_STORE);
+
+        log::info!(target: TARGET,"Latest processed block: {}", latest_processed_block);
+
+        // Now we have fetch missing blocks using previous block hash until we hit latest processed block.
+        // If it's first run, we have to backtrack for BLOCK_AMOUNT_TO_STORE blocks.
+        let mut blocks_to_process =
+            Vec::with_capacity((finalized_block.number - latest_processed_block) as usize);
+
+        let mut current_block = finalized_block.number - 1;
+        let mut prev_block_hash = finalized_block.parent_hash;
+        let block = self
+            .block_rpc
+            .get_block(finalized_block.hash)
+            .await?
+            .ok_or_else(|| eyre::eyre!("Block not found"))?;
```

Comment:
We should query the database first before doing this call.

Reply:
query for what?